.github/scripts/ci.py

```diff
@@ -0,0 +1,346 @@
+        try:
+            out = yaml.safe_load(f)
+        except yaml.YAMLError as e:
+            print_stderr(f"Failed to parse rendered docker-compose file [{e}]")
+            with open(f"{app_dir}/templates/rendered/docker-compose.yaml", "r") as f:
+                print_stderr(f"Syntax Error in rendered docker-compose file:\n{f.read()}")
+            sys.exit(1)
+
+        if args["render_only_debug"]:
+            print_stderr("Successfully rendered docker-compose file:")
+            print_stdout(yaml.dump(out))
+            sys.exit(0)
+
+    print_stderr("Done rendering docker-compose file")
+
+
+def print_docker_compose_config():
+    print_stderr("Printing docker compose config (parsed compose)")
+    cmd = f"{get_base_cmd()} config"
+    print_cmd(cmd)
+    separator_start()
+    res = subprocess.run(cmd, shell=True, capture_output=True)
+    separator_end()
+    if res.returncode != 0:
+        print_stderr("Failed to print docker compose config")
+        if res.stdout:
+            print_stderr(res.stdout.decode("utf-8"))
+        if res.stderr:
+            print_stderr(res.stderr.decode("utf-8"))
+        sys.exit(1)
+
+    if args["render_only"]:
+        print_stdout(res.stdout.decode("utf-8"))
+        sys.exit(0)
+
+    print_stderr(res.stdout.decode("utf-8"))
+
+
+def separator_start():
+    print_stderr("=" * 40 + "+++++" + "=" * 40)
+
+
+def separator_end():
+    print_stderr("=" * 40 + "-----" + "=" * 40)
+
+
+def print_cmd(cmd):
+    print_stderr(f"Running command [{cmd}]")
+
+
+def docker_cleanup():
+    cmd = f"{get_base_cmd()} down --remove-orphans --volumes"
+    print_cmd(cmd)
+    separator_start()
+    subprocess.run(cmd, shell=True)
+    separator_end()
+
+    cmd = f"{get_base_cmd()} rm --force --stop --volumes"
+    print_cmd(cmd)
+    separator_start()
+    subprocess.run(cmd, shell=True)
+    separator_end()
+
+
+def print_logs():
+    cmd = f"{get_base_cmd()} logs"
+    print_cmd(cmd)
+    separator_start()
+    subprocess.run(cmd, shell=True)
+    separator_end()
+
+
+def print_docker_processes():
+    cmd = f"{get_base_cmd()} ps --all"
+    print_cmd(cmd)
+    separator_start()
+    subprocess.run(cmd, shell=True)
+    separator_end()
+
+
+def get_failed_containers():
+    # Outputs one container per line, in json format
+    cmd = f"{get_base_cmd()} ps --all --format json"
+    print_cmd(cmd)
+    failed = subprocess.run(cmd, shell=True, capture_output=True).stdout.decode("utf-8")
+    failed_containers = []
+    for line in failed.split("\n"):
+        if not line:
+            continue
+        try:
+            failed_containers.append(json.loads(line))
+        except json.JSONDecodeError:
+            print_stderr(f"Failed to parse container status output:\n {line}")
+            sys.exit(1)
+
+    actual_failed = []
+    for container in failed_containers:
+        # Skip containers that are exited with 0 (eg init containers), but not restarting (during a restart exit code is 0)
+        if all(
```

Comment:
Why use `all` ?

Reply:
So, to mark a container healthy we need to considering the following:

1. If its exited, the exit code is 0 and the healthcheck is empty. (example here is an init container that runs sometihng and then exits.)
2. It's state is not restarting. (At this point the Health state can have a misleading value)
3. And if there is a healthcheck, the state is healthy.

If any of the above is not true, we consider the container unhealthy.

Reply:
There is no need to use `all` as we are not iterating over anything?
We can just do `if c[cond1] and c[cond2] and c[cond3]` ?

Reply:
If you want to chain conditions you can do
```
if (cond1 or cond1 != something) and (cond2 == 'somethingelse' or cond2 == cond1):
```

Reply:
Fixed!